### ISSUE TITLE
Dart fails with class A<X extends FutureOr<A<X>>>

### ISSUE THREAD

**Issue by iarkh** [dart-lang/sdk#34948]
Dart SDK Version: 2.1.0-dev.8.0
OS: Windows 10 (64 bit)

This is a regression in dev.8.0, dart passes with the same code sample with dev.7.1.

Please note that there is an opened bug #34264 regarding dartanalyzer.

Here is a sample code example:
``` dart
import "dart:async";
class A<X extends FutureOr<A<X>>> {}
main() {}
```

Still analyzer fails here because of #34264.

But in the recent dart version dart started to fail here too, sample output is:
> test.dart:2:9: Error: Type argument '#lib1::A::X' violates the corresponding type variable bound of 'A'.
> Try changing type arguments so that they conform to the bounds.
> class A<X extends FutureOr<A<X>>> {}
>         ^
> test.dart:2:9: Context: Bound of this variable is violated.
> class A<X extends FutureOr<A<X>>> {}
>         ^
> test.dart:1: Error: Type argument '#lib1::A::X' violates the corresponding type variable bound of 'A' in the return type.
> Try changing type arguments so that they conform to the bounds.
> test.dart:2:9: Context: Bound of this variable is violated.
> class A<X extends FutureOr<A<X>>> {}
>         ^
>

**Comment by iarkh** [comment]
Seems like the following code fails with dart because of the same reason:
``` dart
import "dart:async";
class A<X extends A<X>> {}
class B<X extends A<A<X>>> {}
main() {}
```
Sample output is:

> test.dart:3:9: Error: Type argument '#lib1::A<#lib1::B::X>' violates the corresponding type variable bound of 'A'.
> Try changing type arguments so that they conform to the bounds.
> class B<X extends A<A<X>>> {}
>         ^
> test.dart:2:9: Context: Bound of this variable is violated.
> class A<X extends A<X>> {}
>         ^
> test.dart:3:9: Error: Type argument '#lib1::B::X' violates the corresponding type variable bound of 'A'.
> Try changing type arguments so that they conform to the bounds.
> class B<X extends A<A<X>>> {}
>         ^
> test.dart:2:9: Context: Bound of this variable is violated.
> class A<X extends A<X>> {}
>         ^

So please let me know if I should file a separate issue about this.

**Comment by vsmenon** [comment]
Filing on language to clarify behavior and/or whether this is an intended breaking change.

**Comment by iarkh** [comment]
This bug is still reproducible with the recent dart version:
Dart SDK version: 2.11.0-176.0.dev (dev) (Mon Sep 28 19:05:52 2020 -0700) on "windows_x64"

**Comment by scheglov** [comment]
The fix landed as 499bae02e06e8787a6751b4a961577363f33e475.

It seems that this change should be in `2.11.0-176.0.dev`. When I try with the bleeding edge build, it seems to work.
```
scheglov@scheglov-macbookpro2:~/Source/Dart/sdk.git/sdk (master)$ cat /Users/scheglov/dart/test/bin/test.dart
import "dart:async";

class B<X extends Future<A<X>>> {}
class A<X extends FutureOr<A<X>>> {}

A a = 0;
B b = 0;
scheglov@scheglov-macbookpro2:~/Source/Dart/sdk.git/sdk (master)$ dartanalyzer --version
dartanalyzer version 2.11.0-edge.919f68d8ac2da1f1498bceb3cda610b4b0ce9905
scheglov@scheglov-macbookpro2:~/Source/Dart/sdk.git/sdk (master)$ dartanalyzer /Users/scheglov/dart/test/bin/test.dart
Analyzing /Users/scheglov/dart/test/bin/test.dart...
  error • A value of type 'int' can't be assigned to a variable of type 'A<FutureOr<A<dynamic>>>'. • /Users/scheglov/dart/test/bin/test.dart:6:7 • invalid_assignment
  error • A value of type 'int' can't be assigned to a variable of type 'B<Future<A<dynamic>>>'. • /Users/scheglov/dart/test/bin/test.dart:7:7 • invalid_assignment
2 errors found.
```

**Comment by eernstg** [comment]
Closing: The example program is accepted by `dart` and `dartanalyzer` as of a49217bc3aa46e5406ad72017901490c2913c409.
